driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -118,4 +119,36 @@ public <R extends Expression> R cond(final R left, final R right) {
         return newMqlExpression(ast("$cond", left, right));
     }
 
+
+    /** @see ArrayExpression */
+
+    @Override
+    public <R extends Expression> ArrayExpression<R> map(final Function<T, ? extends R> in) {
+        T varThis = variable("$$this");
+        return new MqlExpression<>((cr) -> astDoc("$map", new BsonDocument()
```

Comment:
I am using this method to mark where we would need to change the implementation (to be an AST). This makes the task of contemplating an AST implementation easier.

Reply:
Shouldn't such tracking be done via a type system? That way it would really track things by forcing you to create an instance of the right type (AST) in places where it's needed. The constructor of `MqlExpression` appears to be taking an AST, which is what I asked about in the previous PR. I did not understand the answer there and just resolved the concern, but here we come back to that. We should probably have a call about this.

Regardless of the above, the implementation of `astDoc` appears to need a change: it should just create a new `BsonDocument` instead of creating a function that is applied right away.

Reply:
Done in https://github.com/mongodb/mongo-java-driver/pull/1037/files